api/auth.proto

```diff
@@ -44,3 +46,78 @@ message AuthAction {
   // List of rules
   repeated Rule rule = 2;
 }
+
+service Authorization {
+	rpc Check(CheckRequest) returns (CheckResponse);
+}
+
+message CheckRequest {
+
+	// Address of the downstream source of the request.
+	Address source = 1;
+
+	// Address of the upstream destination of the request.
+	Address destination = 2; 
+
+	// If the request is an HTTP request, the parameters
+	// of the HTTP request.
+	HTTPRequest http = 3;
+
+	// If the request came over a secure TLS connection, information about the connection and 
+	// downstream client.
+	TLSConnection tls = 4;
+}
+
+message HTTPRequest {
+
+	// HTTP Version of the request, e.g. "1.1"
+	string version = 1;
+
+	// Method, e.g. "GET"
+	string method = 2;
+
+	// The value of the Host: header
+	string host = 3;
+
+	// The request URI path.
+	string path = 4;
+
+	// Additional HTTP headers and their values.
+	map<string, string> headers = 5;
+}
+
+message TLSConnection {
+
+	// True if the connection is encrypted with TLS.  False otherwise. 
+	bool tls_encrypted = 1;
+
+	// If the client presented an X.509 certificate, it is contained here.
+	X509Cert client_cert = 2;
+
+	// The server certificate used to secure the connection.
+	X509Cert server_cert = 3;
+}
+
+
+// An X.509 certificate, with any corresponding fields already decoded by Envoy included.
+message X509Cert {
```

Comment:
Can we make use of https://github.com/envoyproxy/data-plane-api/blob/master/api/sds.proto#L72 here?

Reply:
+1

Reply:
Reusing that message is awkward, because it is designed to express the set of acceptable certificates, and what we want is *the actual certificate* that was accepted.  If we merged them we'd reuse a few fields, but other fields would be ignored depending on whether it was used as a validation context or an authorization context.  I think that makes documentation tricky and invites operator confusion, so isn't worth the benefit.

Reply:
So, we have similar info in 3 places if we add this; `X509Cert` here, `CertificateValidationContext` and also in the XFCC header (https://github.com/envoyproxy/data-plane-api/blob/master/api/filter/http/http_connection_manager.proto#L141 and https://www.envoyproxy.io/docs/envoy/latest/configuration/http_conn_man/headers.html#config-http-conn-man-headers-x-forwarded-client-cert).

I almost wish we had made XFCC a JSON/proto encoded field.. I wonder for external auth validation, whether what you want is essentially the same as the info that we're jamming into the XFCC header.

I agree with your above observation @spikecurtis. Could you spend a few mins looking at XFCC (which is essentially a header we invented) and looking to see if there is any meaningful overlap between what we want for auth filter and this?

Reply:
@htuch The XFCC header is indeed really similar.  I'm not crazy about reusing the string encoding of the XFCC here because it will mean the service receiving this message has to implement additional decoding logic.  If it's all in proto, it means the proto library automatically handles decoding.  If it were complex or widely used, I could see the value in reusing the encoding, but it's only a couple fields.

I'm not in a great position to comment on how desirable it would be to change XFCC to be proto encoded at this point. I'm assuming there are applications in Lyft that consume this header?

Reply:
@mandarjog We are experimenting with asserting additional identity claims on the X.509 certificate.  Claims like group membership or labels don't lend themselves well to being encoded in SANs.  Passing the raw certificate is coarse-grained, but allows this kind of experimentation without rewriting the API.

I'm not wed to this mechanism, though, so we could discuss others if you'd like.  Some other ideas I've had:
 
1. Envoy could pass any X.509 extensions that it doesn't understand
2. We could design & standardize an extensible claims encoding

I don't really favor the effort of option 2 before we and others start shipping products that use this mechanism and have a chance to validate it with end users.